### ISSUE TITLE
Nested templates error

### ISSUE THREAD

**Issue by tttapa** [michaeljones/breathe#728]
Breathe + Sphinx fails when parsing nested templates, e.g.
```cpp
template <template <class> class TT>
class C {};
```
Doxygen generates the following XML:
```xml
    <templateparamlist>
      <param>
        <type>template&lt; class &gt; class</type>
        <declname>TT</declname>
        <defname>TT</defname>
      </param>
    </templateparamlist>
```
Running Sphinx causes the following error:
```text
Exception occurred:
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 275, in fail
    raise self._make_multi_error(errors, '')
sphinx.util.cfamily.DefinitionError: Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 8]
  template< class > class
  --------^
```

<details>
<summary><b>Full backtrace</b></summary>

```text
# Sphinx version: 4.1.2
# Python version: 3.9.6 (CPython)
# Docutils version: 0.17.1 release
# Jinja2 version: 3.0.1
# Last messages:
#   Running Sphinx v4.1.2
#   building [mo]: targets for 0 po files that are out of date
#   building [html]: targets for 1 source files that are out of date
#   updating environment:
#   [new config]
#   1 added, 0 changed, 0 removed
#   reading sources... [100%] index
# Loaded extensions:
#   sphinx.ext.mathjax (4.1.2) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (2.0.0) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.5) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/alabaster/__init__.py
#   breathe (4.30.0) from /home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/__init__.py
Traceback (most recent call last):
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/application.py", line 343, in build
    self.builder.build_update()
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 307, in build
    updated_docnames = set(self.read())
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 414, in read
    self._read_serial(docnames)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 435, in _read_serial
    self.read_doc(docname)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 475, in read_doc
    doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/io.py", line 188, in read_doc
    pub.publish()
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/core.py", line 217, in publish
    self.document = self.reader.read(self.source, self.parser,
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/io.py", line 108, in read
    self.parse()
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/readers/__init__.py", line 78, in parse
    self.parser.parse(self.input, document)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/parsers.py", line 100, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 170, in run
    results = StateMachineWS.run(self, input_lines, input_offset,
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/statemachine.py", line 239, in run
    context, next_state, result = self.check_line(
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/statemachine.py", line 451, in check_line
    return method(match, context, next_state)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2769, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 393, in new_subsection
    newabsoffset = self.nested_parse(
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 281, in nested_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/statemachine.py", line 239, in run
    context, next_state, result = self.check_line(
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/statemachine.py", line 451, in check_line
    return method(match, context, next_state)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2342, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2354, in explicit_construct
    return method(self, expmatch)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2096, in directive
    return self.run_directive(
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2146, in run_directive
    result = directive_instance.run()
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/directives.py", line 292, in run
    return self.render(matches[0], project_info, filter_, target_handler, mask_factory,
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/directive/base.py", line 96, in render
    return object_renderer.render(node_stack[0], context)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2279, in render
    result = method(self, node)  # type: ignore
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2166, in dispatch_compound
    return self.visit_compound(node)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 1063, in visit_compound
    return self.visit_class(node)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 977, in visit_class
    self.create_template_prefix(nodeDef),
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 780, in create_template_prefix
    nodes = self.render(decl.templateparamlist)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2279, in render
    result = method(self, node)  # type: ignore
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2089, in visit_templateparamlist
    nodelist.extend(self.visit_param(item, insertDeclNameByParsing=True))
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2051, in visit_param
    ast = parser._parse_type(named='single', outer='templateParam')
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 6415, in _parse_type
    declSpecs = self._parse_decl_specs(outer=outer)
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 6109, in _parse_decl_specs
    trailing = self._parse_trailing_type_spec()
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 5874, in _parse_trailing_type_spec
    nestedName = self._parse_nested_name()
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 5800, in _parse_nested_name
    self.fail("Expected identifier in nested name, "
  File "/home/pieter/Random/Cpp/Breathe-bug/venv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 275, in fail
    raise self._make_multi_error(errors, '')
sphinx.util.cfamily.DefinitionError: Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 8]
  template< class > class
  --------^
```

</details>

The issue is caused by the call to `sphinx.DefinitionParser._parse_type()` here:

https://github.com/michaeljones/breathe/blob/d8add71c4613f6073928ad221400241407c17eb3/breathe/renderer/sphinxrenderer.py#L2089

The `_parse_type()` function is unable to parse the string `template< class > class`.

ZIP file with the source and config files to reproduce the issue is attached:

[**Breathe-bug.zip**](https://github.com/michaeljones/breathe/files/7042289/Breathe-bug.zip)

On Linux you should be able to just run the `reproduce.sh` bash script to reproduce the issue.

---

**Edit**: Variadic templates don't work either:

```cpp
/// The following is a variadic template.
template <class... Args>
class C {};
```
```xml
    <templateparamlist>
      <param>
        <type>class...</type>
        <declname>Args</declname>
        <defname>Args</defname>
      </param>
    </templateparamlist>
```
```text
Exception occurred:
  File "/home/pieter/Random/Cpp/Breathe-bug2/venv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 275, in fail
    raise self._make_multi_error(errors, '')
sphinx.util.cfamily.DefinitionError: Invalid C++ declaration: Expected identifier in nested name. [error at 5]
  class...
  -----^
```

**Comment by jakobandersen** [comment]
Thanks for the report. This should be fixed by #711.

**Comment by vermeeren** [comment]
Assuming fixed via #711.
